
# Introduction to Numpy
    
<img src='images/Numpy.jpg' alt = "Numpy" style="width:600; height:300"/>    


## Objectives

In this section of the course, students will learn how to import Numpy library, create both one and two dimensional array and do various operations with it. At the end of the Numpy section, students will be able to work and maniputate n dimensional array and be able to work with linear algebra and solve some simple system of linear equations.

## Table of content

* Introduction to Numpy
* Installation instruction guide
* How to import Numpy library
* Python Lists and NumPy Arrays
* Array Slicing
* Numpy Attributes
* Other ways to create arrays
* Multi dimensional array
* Systems of Linear Equations

# Introduction to Numpy
NumPy is a Python package/librarry that stands for `Numerical Python`. It is the core library for scientific computing, which contains a powerful n-dimensional array object. It is also a linear algebra for python and almost all of the libraries in the Python ecosystem rely on it as one of their main building blocks. It is incredibly fast, as it has bindings to C libraries.

## **Prerequisites**

A basic understanding of Python programming in CS14 (Programming in Python) is required. 

## **Installation Instruction guide**

If you installed the [Anaconda distribution](https://www.anaconda.com/products/individual) of Python - it includes Python, NumPy, and other commonly used packages for scientific computing and data science.Therefore, no further installation steps are necessary. We recommend you use the [Anaconda distribution](https://www.anaconda.com/products/individual) of Python as you begin your data science journey.

If you use a version of Python from python.org or a version of Python that came with your operating system, the **Anaconda Prompt** and **conda** or **pip** can be used to install NumPy.

### **Install NumPy with the Anaconda Prompt**

To install NumPy, open the Anaconda Prompt and type:

`conda install numpy`

Type y for yes when prompted.

### **Install NumPy with pip**

To install NumPy with pip, bring up a terminal window and type:


$ pip install numpy

This command installs NumPy in the current working Python environment.

## Importing the NumPy module
There are several ways to import NumPy. The standard approach is to use a simple import statement:


In [1]:
import numpy

However, for large amounts of calls to NumPy functions, it can become tedious to write
numpy.X over and over again. Instead, it is common to import under the briefer name np:

Numpy has many built-in functions and to use each of the functions, we will need to call them from numpy. For example
numpy.array, numpy.mean, numpy.log10, etc. Here, we see that we are calling numpy.something over and over again. Instead, it is common in Python to use alias. For example:

In [2]:
import numpy as np

## Python Lists and NumPy Arrays

This section introduces NumPy arrays then explains the difference between Python lists and NumPy arrays. NumPy is used to construct homogeneous arrays and perform mathematical operations on arrays. A NumPy array is different from a Python list. The data types stored in a Python list can all be different.


In [3]:
python_list = ["Ethiopia", 5, 17.9, True]

type(python_list)

list

The Python list above contains four different data types:
- "Ethiopia" is a string
- 5 is an integer
- 17.9 is a float, 
- True is a boolean.

## What is Python Numpy Array?
NumPy arrays are a bit like Python lists, but still very much different at the same time. The simplest way to create an array in Numpy is to use Python List.

In [1]:
myPythonList = [1, 4, 7, 2, 12, 17]

In [2]:
name = ["Joke", "Jamal", "Sadiq"]

In [3]:
name

['Joke', 'Jamal', 'Sadiq']

We can convert python list to a numpy array by using the object np.array.

In [5]:
numpy_array_from_list = np.array(myPythonList)

type(numpy_array_from_list)

numpy.ndarray

**To display the output**:

In [6]:
numpy_array_from_list

array([ 1,  4,  7,  2, 12, 17])

In practice, there is no need to declare a Python List. The operation can be combined.

In [7]:
new_array  = np.array([1, 9, 8, 3, 12])
new_array

array([ 1,  9,  8,  3, 12])

## Mathematical operations on one dimensional array
You can perform mathematical operations like additions, subtraction, division, and multiplication on an array. The syntax is the array name followed by the operation (+, -, *, /).

## Example 1

In [8]:
a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

# Addition
a + 2

array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [9]:
# Subtraction
a - 3

array([-2, -1,  0,  1,  2,  3,  4,  5,  6,  7])

In [10]:
# Multiplication
a * 2

array([ 2,  4,  6,  8, 10, 12, 14, 16, 18, 20])

In [11]:
# Division
a/2

array([0.5, 1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. ])

In [12]:
# Exponentiation

a**3

array([   1,    8,   27,   64,  125,  216,  343,  512,  729, 1000],
      dtype=int32)

## Example 2

When standard mathematical operations are used with arrays, they are applied on an element-by-element basis. This means that the arrays should be the same size during addition,
subtraction, etc.

In [13]:
a = np.array([5, 2, 7, 4, 12])

b = np.array([3, 2, 4, 1, 6])

In [14]:
print(a + b)

[ 8  4 11  5 18]


In [15]:
print(a - b)

[2 0 3 3 6]


In [16]:
print(a * b)

[15  4 28  4 72]


In [17]:
print(b / a)


[0.6        1.         0.57142857 0.25       0.5       ]


In [18]:
print(a % b) # Remainder when you divide each of the elements in a by b

[2 0 3 0 0]


In [19]:
b**a

array([        243,           4,       16384,           1, -2118184960],
      dtype=int32)

In [20]:
np.sqrt(a) # square root of each of the elements in a

array([2.23606798, 1.41421356, 2.64575131, 2.        , 3.46410162])

# Class activity 1 (Peer to peer review activity)

Convert the following Python lists to Numpy arrays:

even_list = [2, 4, 6, 8, 10]

odd_list = [1, 3, 5, 7, 9]

and add the resulting arrays together and name it even_odd_array

# One-dimensional Array Indexing

Array elements are accessed, sliced, and manipulated just like lists. 

In [21]:
a = np.array([1, 2, -1, 4, -5, 6, 7, 0, 9, 10])

print(a)

[ 1  2 -1  4 -5  6  7  0  9 10]


**Remember counting in Python starts at 0 and ends at n-1**.

The index (or location) of each value in the array is shown below:

The value 1 has an index of 0. We could also say 1 is in location 0 of the array. The value 4 has an index of 3 and the value 8 has an index of 9. 

In [22]:
a[0] # accessing the first element in array a

1

In [23]:
a[3] # accessing the third element in array a

4

In [24]:
a[2] = 5 # accessing the first element in array a and replacing the result with 5
a

array([ 1,  2,  5,  4, -5,  6,  7,  0,  9, 10])

# Class activity 2 (Peer to peer discusion)

How will you access the $3^{rd}$ element in the 

`age_array = np.array([19, 17, 15, 13, 20, 11, 18, 10, 17])`

# Array Slicing

Values stored within an array can be accessed simultaneously with array slicing. To pull out a section or slice of an array, the colon operator : is used when calling the index. The general form is:

`array [start : stop]`

The index of the slice is specified in `[start : stop]`. Remember Python counting starts at $0$ and ends at $n-1$. The index `[0 : 2]` pulls the first two values out of an array. The index `[1 : 3]` pulls the second and third values out of an array.

# Example 1

In [25]:
a = np.array([2, 4, 6, 5, 8, 9])
print(a)

[2 4 6 5 8 9]


In [26]:
b = a[0:2]
print(b)

[2 4]


## Example 2

In [27]:
x = np.array([5, 8, 9, 2, 4, 6, 5])
print(x)

[5 8 9 2 4 6 5]


In [28]:
print(x[1:3])

[8 9]


On either sides of the colon, a blank stands for "default".

- `[:2]` corresponds to `[start=default:stop=2]`
- `[1:]` corresponds to `[start=1:stop=default]`

Therefore, the slicing operation `[:2]` pulls out the first and second values in an array. The slicing operation `[1:]` pull out the second through the last values in an array. The examples below illustrate the default stop value is the last value in the array.

## Example 4

In [29]:
a = np.array([2, 4, 6, 8, 10, 0, 1, 5])
print(a)

[ 2  4  6  8 10  0  1  5]


In [30]:
b = a[1:]
print(b)

[ 4  6  8 10  0  1  5]


## Example 5

In [31]:
x = np.array([0, 5, 6, 1, 0, 8, 1, 5, 9])
print(x)

[0 5 6 1 0 8 1 5 9]


In [32]:
y = x[2:]
print(y)

[6 1 0 8 1 5 9]


The next examples shows the default `start` value is the first value in the array.

In [33]:
a = np.array([2, 4, 6, 8, 1, 3, 6])
print(a)

[2 4 6 8 1 3 6]


In [34]:
b = a[:3]
print(b)

[2 4 6]


The following indexing operations output the same array.

In [35]:
a = np.array([2, 1, 7, 4, 6, 8, 3])

b = a[0:7] # [start=0:stop= 6]

print(b)

c = a[:7] # [start=0:stop= 6]

print(c)

d = a[0:] # [start=0:stop= 6]

print(d)

e = a[:] # [start=0:stop= 6]

print(e)

[2 1 7 4 6 8 3]
[2 1 7 4 6 8 3]
[2 1 7 4 6 8 3]
[2 1 7 4 6 8 3]


# Class activity 3 (Pilot question 1)

Consider:

`a = np.array([2, 1, 0, 4, 7, 4, 6, 8, 3])`

if you want to pull out the first three values in `a`, that is, to look like 

`array([2, 1, 0])`

what will you do?

# Pilot answer 1

`a[:3]`

# Numpy Attributes

Array attributes reflect information that is intrinsic to the array itself. Generally, accessing an array through its attributes allows you to get the intrinsic properties of the array. Some commonly used attributes are:

* shape:    indicates the dimension of an array
* size:     returns the total number of elements in the array
* dtype:    returns the type of elements in the array, i.e., int64, character


You can check the shape of the array with the object shape preceded by the name of the array. In the same way, you can check the type with dtypes.


## Example 1

In [36]:
even_array  = np.array([2, 4, 6, 8, 10, 12, 14, 16])

even_array

array([ 2,  4,  6,  8, 10, 12, 14, 16])

In [37]:
# dimension of the array

print(even_array.shape)

(8,)


In [38]:
# total element in the array
print(even_array.size)

8


In [39]:
# datatype in the array
print(even_array.dtype) 

int32


## Example 2

In [40]:
some_array = np.array([12, 62, 65,  7, 21, 60, 10, 87, 14, 43, 51, 10, 38, 95, 26, 11, 46, 47, 34, 68, 58, 77, 13, 10, 45])

some_array

array([12, 62, 65,  7, 21, 60, 10, 87, 14, 43, 51, 10, 38, 95, 26, 11, 46,
       47, 34, 68, 58, 77, 13, 10, 45])

In [41]:
some_array.shape

(25,)

In [42]:
some_array.size

25

# Other ways to create arrays

### arange
The arange function is similar to the range function but returns an array:

In [43]:
import numpy as np
np.arange(5)

array([0, 1, 2, 3, 4])

In [44]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [45]:
np.arange(1, 10, 2)

array([1, 3, 5, 7, 9])

### linspace
Return evenly spaced numbers over a specified interval.

In [46]:
np.linspace(0,10,3)

array([ 0.,  5., 10.])

In [47]:
np.linspace(0, 50, 10)

array([ 0.        ,  5.55555556, 11.11111111, 16.66666667, 22.22222222,
       27.77777778, 33.33333333, 38.88888889, 44.44444444, 50.        ])

## random

Numpy also has lots of ways to create random number arrays:

### rand

`numpy.random.rand()` creates an array of the given shape and populate it with random samples from a uniform distribution over [0, 1).

In [48]:
np.random.rand(3)

array([0.25576206, 0.4827032 , 0.27154667])

In [49]:
np.random.rand(5, 5)

array([[0.89113549, 0.40135573, 0.92106703, 0.09391136, 0.36695159],
       [0.74195433, 0.79097947, 0.19500484, 0.94726795, 0.94544754],
       [0.42514933, 0.11020008, 0.00662118, 0.29685049, 0.16650326],
       [0.40384536, 0.25307622, 0.70273094, 0.06005833, 0.6445571 ],
       [0.72601159, 0.69857904, 0.32565753, 0.80996799, 0.0766876 ]])

### randn

`numpy.random.randn()` returns a sample (or samples) from the "standard normal" distribution. Unlike `rand` which is uniform:

In [50]:
np.random.randn(2)

array([-0.10319486,  1.42912385])

In [51]:
np.random.randn(50)

array([-3.71617381e-01,  4.99634720e-01, -1.55403861e+00, -2.06997183e+00,
        9.32990841e-01, -1.81924815e+00,  5.75458269e-01, -5.35094662e-01,
       -1.46386148e+00,  6.69068357e-01,  6.96823466e-01, -1.84120177e+00,
       -7.71288146e-01,  1.03722825e+00, -4.27671338e-01,  1.51695913e-02,
        5.29912627e-01,  1.85208439e-01,  3.36423416e-01, -1.35622915e-01,
        2.64381160e-01,  8.71614526e-01,  1.56528186e+00,  2.68996315e-01,
        6.84663796e-01,  1.35351647e+00, -5.93678195e-01, -5.41750533e-01,
        4.76491057e-01, -2.00696350e+00, -7.90785937e-01,  2.23476052e-01,
       -1.19757240e+00,  7.12481625e-01,  1.76254776e+00, -4.00040328e-02,
        3.72621459e-01,  9.30932412e-01,  2.44406692e+00, -8.93453111e-01,
       -1.15249838e+00, -1.32610761e+00,  6.50563964e-01, -5.15101903e-01,
       -4.22744240e-01, -2.05050950e+00, -4.49072927e-01, -1.43982102e-03,
        2.09279770e+00,  1.33514311e-01])

In [52]:
np.random.randn(5, 5)

array([[-1.23459923,  2.56913157, -2.38803784, -0.20361266, -0.85192003],
       [ 0.27303167, -0.80724834,  0.248172  ,  0.39040131, -0.27633046],
       [ 1.16522723, -1.12107195, -1.32237583,  0.55013136,  0.12333403],
       [-0.52629067, -0.42857803,  0.00396803,  0.37131638,  0.0883102 ],
       [-0.15919958, -0.33084369, -0.24295596, -0.2678193 ,  0.26894644]])

### randint
Return random integers from low (inclusive) to high (exclusive).

In [53]:
np.random.randint(1, 100)

93

In [54]:
np.random.randint(1, 100, 10)

array([73, 76, 93, 19,  8, 59, 65, 45, 69, 93])

# Class activity 4

Create $100$ random integers from $20$ to $90$ using `np.random.randint()` function.

# Multi dimensional array

Arrays can be multidimensional. Unlike lists, different axes are accessed using commas inside bracket notation. A simple 2-D array is defined by a list of lists. Here is an example with a two-dimensional array (e.g. a matrix).

## Example 1

$A = 
\begin{pmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{pmatrix}$


In [55]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

print(A)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [56]:
print(A.shape) # dimension of the array. This returns a tupple

(3, 3)


In [57]:
print(A.size)# Total elements in the array

9


## Example 2

$another\_array = 
\begin{pmatrix}
5 & 6 & 7 \\
4 & 2 & 1 \\
3 & 7 & 1
\end{pmatrix}$

In [58]:
another_array = np.array([[5, 6, 7], [4, 2, 1], [3, 7, 1]])

print(another_array)

[[5 6 7]
 [4 2 1]
 [3 7 1]]


In [59]:
print(another_array.shape) 

(3, 3)


In [60]:
print(another_array.size)

9


# Class activity 5

Consider:

$x = 
\begin{pmatrix}
0 & 6 &  3\\
5 & 2 & 7 \\
3 & 4 & 1
\end{pmatrix}$

1. Develop a two dimensional array for x using numpy

2. What is the shape of x?

3. What is the number of elements in x?

## Two-Dimensional Array Indexing

Values in a 2-D array can be accessed using the general notation below:

`value = array name [row, col]`

Where **value** is the value pulled out of the 2-D array and `[row, col]` specifies the row and column index of the value. Remember Python counting starts at 0, so the first row is row zero and the first column is column zero. We can access the value 2 in the array above by calling the row and column index `[0, 1]`. This corresponds to the 1st row (remember row 0 is the first row) and the 2nd column (column 0 is the first column).


$A = 
\begin{matrix}
C0& C1 & C2\\
1 & 2 & 3 & R0 \\
4 & 5 & 6 & R1 \\
7 & 8 & 9& R 2
\end{matrix}$


## Example 1

In [61]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

print(A)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [62]:
A[0, 1] # value 2 is is row 0 and column 1

2

In [63]:
A[1, 0] # value 4 is is row 1 and column 0

4

## Example 2

In [64]:
a = np.array([[2,3,4],[6,7,8]])
print(a)

[[2 3 4]
 [6 7 8]]


In [65]:
a[1, 2]

8

In [66]:
a[1, 2] = 20

In [67]:
print(a)

[[ 2  3  4]
 [ 6  7 20]]


## Other array indexing

2D NumPy arrays can also be sliced with the general form:

`array[row = start_row:end_row, col = start_col:end_col]`

The code section below creates a two row by four column array and indexes out the first two rows and the first three columns.

In [68]:
a = np.array([[2, 4, 6, 8], [0, 5, 4, 1]])
print(a)

[[2 4 6 8]
 [0 5 4 1]]


In [69]:
a.shape

(2, 4)

In [70]:
b = a[0:2, 0:3]
print(b)

[[2 4 6]
 [0 5 4]]


The code section below slices out the first two rows and all columns from array a.

In [71]:
b = a[:2, :]  #[first two rows, all columns]
print(b)

[[2 4 6 8]
 [0 5 4 1]]


Again, a blank represents defaults the first index or the last index. The colon operator all by itself also represents "all" (default start: default stop).

In [72]:
b = a[:, :]  #[all rows, all columns]
print(b)

[[2 4 6 8]
 [0 5 4 1]]


## 2D Array mathematics
When standard mathematical operations are used with arrays, they are applied on an element-by-element basis. This means that the arrays should be the same size during addition,
subtraction, etc.:

## Example 1

In [73]:
a = np.array([[3, 8], [4, 6]])
print(a)

[[3 8]
 [4 6]]


In [74]:
b =  np.array([[4, 0], [1, -9]])
print(b)

[[ 4  0]
 [ 1 -9]]


In [75]:
# Addition
a + b 

array([[ 7,  8],
       [ 5, -3]])

In [76]:
# Subtraction 
a - b

array([[-1,  8],
       [ 3, 15]])

In [77]:
# Multiply by a Constant

a*2

array([[ 6, 16],
       [ 8, 12]])

**Multiplying by Another Matrix**

Multiplication of two matrices is possible only when number of columns in first matrix equals number of rows in second matrix. Multiplication by another matrix uses a `dot product` or with `@` operator

In [78]:
np.dot(a, b)

array([[ 20, -72],
       [ 22, -54]])

In [79]:
a @ b

array([[ 20, -72],
       [ 22, -54]])

# Class activity 6

Consider the following numpy arrays:

`a = np.array([[3, 1, 8], [4,2, 6], [3, 7, 4]])`

`b =  np.array([[4, 0, 6], [1, -9, 2], [1, 2, 3]])`


Find the value of :

1. `a - b`

2. `a + b`

3. `a` $\times$ `b`

4. What is result of `a[0 : 1, 1 : 2]`?

# NumPy use cases: Systems of Linear Equations

Our knowledge in numpy array can be used to solve system of linear equations. Remember in your junior high school, you were taught how to solve simultaneous equation. The terms simultaneous equations or systems of linear equations refer to conditions where two or more unknown variables are related to each other through an equal number of equations.


A system of linear equations is shown below:
    
    
$2x + 3y = 8$
    
$x - y = -1$

We have two unknowns variables x and y and two equations i.e. $2x + 3y = 8$ and $x - y = -1$ .

## numpy.linalg.solve()

NumPy's `np.linalg.solve()` function can be used to solve this system of equations for the variables x and y.

## Steps to follow:

The steps to solve the system of linear equations with np.linalg.solve() are below:

1. Create NumPy array `A` as a 2 by 2 array of the coefficients in x and y

2. Create a NumPy array `b` as the right-hand side of the equations

3. Solve for the values of x and y using `np.linalg.solve(A, b)`.

The resulting array has two entries. One entry for each variable.

## Example 1


Solve simulataneuos equation below:
    
    
$2x + 3y = 8$
    
$x - y = -1$

## Solution

We want to solve for the unknown x and y.

In [80]:
import numpy as np

A = np.array([[2, 3], [1, -1]])
b = np.array([8, -1])
x = np.linalg.solve(A, b)
x

array([1., 2.])

`x = x[0]`

`y = x[1]`

Therefore, 
$x = 1$ and $y = 2$.

## Example 2

Solve the system of linear equation:

$2x -y = 3$

$x -3y = -2$

## Solution

We want to solve for the unknown x and y.

In [81]:
import numpy as np

A = np.array([[2, -1], [1, -3]])
b = np.array([3, -2])
x = np.linalg.solve(A, b)
x

array([2.2, 1.4])

$x = x[0] = 2.2$

$y = x[1] = 1.4$

# Class activity 6

The following simultaneous/system of equation:
    
3x + 4y = 24

4x + 3y = 22    
    
has been broken down to numpy array for you:
    

`A = np.array([[3, 4], [4, 3]])`

`b =  np.array([24, 22])`


Complete the following code to solve for x and y: 
    
`np.linalg( _ , _)` 